openerp/osv/fields.py

```diff
@@ -1007,6 +1007,25 @@ def set(self, cr, model, id, name, values, user=None, context=None):
             return
         rel, id1, id2 = self._sql_names(model)
         obj = model.pool[self._obj]
+
+        def insert(ids):
+            # beware of duplicates when inserting
+            query = """ INSERT INTO {rel} ({id1}, {id2})
+                        (SELECT %s, unnest(%s)) EXCEPT (SELECT {id1}, {id2} FROM {rel})
```

Comment:
Indeed, I made a test relation `rel(id1, id2)` with a million rows, and analyzed both ways. The current implementation gives

    =# explain analyze insert into rel(id1, id2) (select 42, unnest(array[0,1,2,3,4,5,6,7,8,9])) except (select id1, id2 from rel);
                                                                         QUERY PLAN
    -----------------------------------------------------------------------------------------------------------------------------------------------------
     Insert on rel  (cost=0.00..29427.02 rows=1 width=8) (actual time=351.102..351.102 rows=0 loops=1)
       ->  Subquery Scan on "*SELECT*"  (cost=0.00..29427.02 rows=1 width=8) (actual time=351.100..351.100 rows=0 loops=1)
             ->  HashSetOp Except  (cost=0.00..29427.01 rows=1 width=8) (actual time=351.098..351.098 rows=0 loops=1)
                   ->  Append  (cost=0.00..24426.51 rows=1000100 width=8) (actual time=0.009..245.442 rows=1000010 loops=1)
                         ->  Subquery Scan on "*SELECT* 1"  (cost=0.00..1.51 rows=100 width=0) (actual time=0.008..0.016 rows=10 loops=1)
                               ->  Result  (cost=0.00..0.51 rows=100 width=0) (actual time=0.007..0.011 rows=10 loops=1)
                         ->  Subquery Scan on "*SELECT* 2"  (cost=0.00..24425.00 rows=1000000 width=8) (actual time=0.013..185.737 rows=1000000 loops=1)
                               ->  Seq Scan on rel rel_1  (cost=0.00..14425.00 rows=1000000 width=8) (actual time=0.012..73.120 rows=1000000 loops=1)
     Total runtime: 351.164 ms
    (9 rows)

Your suggested change gives

    =# explain analyze insert into rel(id1, id2) (select 42, unnest(array[0,1,2,3,4,5,6,7,8,9])) except (select id1, id2 from rel where id1=42);
                                                                         QUERY PLAN
    -----------------------------------------------------------------------------------------------------------------------------------------------------
     Insert on rel  (cost=0.00..2385.35 rows=1 width=8) (actual time=4.054..4.054 rows=0 loops=1)
       ->  Subquery Scan on "*SELECT*"  (cost=0.00..2385.35 rows=1 width=8) (actual time=4.052..4.052 rows=0 loops=1)
             ->  HashSetOp Except  (cost=0.00..2385.34 rows=1 width=7) (actual time=4.051..4.051 rows=0 loops=1)
                   ->  Append  (cost=0.00..2379.99 rows=1070 width=7) (actual time=0.013..3.708 rows=1010 loops=1)
                         ->  Subquery Scan on "*SELECT* 1"  (cost=0.00..1.51 rows=100 width=0) (actual time=0.012..0.021 rows=10 loops=1)
                               ->  Result  (cost=0.00..0.51 rows=100 width=0) (actual time=0.011..0.014 rows=10 loops=1)
                         ->  Subquery Scan on "*SELECT* 2"  (cost=23.94..2378.48 rows=970 width=8) (actual time=0.814..3.488 rows=1000 loops=1)
                               ->  Bitmap Heap Scan on rel rel_1  (cost=23.94..2368.78 rows=970 width=8) (actual time=0.812..3.106 rows=1000 loops=1)
                                     Recheck Cond: (id1 = 42)
                                     ->  Bitmap Index Scan on rel_pkey  (cost=0.00..23.70 rows=970 width=0) (actual time=0.405..0.405 rows=1000 loops=1)
                                           Index Cond: (id1 = 42)
     Total runtime: 4.149 ms
    (12 rows)

The most specific query is clearly the winner :smile:
I will make the change accordingly.